test/test_va_api_fixture.cpp

```diff
@@ -507,6 +638,88 @@ void VAAPIFixture::doCreateConfigToFail(VAProfile profile,
               VA_STATUS_ERROR_INVALID_CONFIG);
 }
 
+void VAAPIFixture::doUploadImage()
+{
+    int box_width_loc=8;
+    int row_shift_loc=0;
+    int i;
+
+    for (i=0; i<SURFACE_NUM; i++) {
+        upload_surface(m_vaDisplay, m_surfaceID[i], box_width_loc, row_shift_loc, 0);
+
+        row_shift_loc++;
+        if (row_shift_loc==(2*box_width_loc)) row_shift_loc= 0;
+    }
+}
+
+void VAAPIFixture::doGetImage(uint32_t currentFmt, VASurfaceID surface_id,
+                                std::pair<uint32_t, uint32_t> currentResolution)
+{
+    doTestCreateImage(currentFmt, currentResolution);
+
+    ASSERT_STATUS(vaGetImage(m_vaDisplay, surface_id, 0, 0,
+                currentResolution.first, currentResolution.second, m_image.image_id));
+}
+
+VASurfaceID VAAPIFixture::doGetNextSurface(int *index)
+{
+    VASurfaceID surface_id = VA_INVALID_SURFACE;
+    int i;
+
+    while (surface_id == VA_INVALID_SURFACE){
+        i = *index;
+        i++;
+        if (i == SURFACE_NUM)
+            i = 0;
+        *index = i;
+        surface_id =  m_surfaceID[i];
+    }
+    return surface_id;
+}
+
+void VAAPIFixture::doPutImage(std::pair<uint32_t, uint32_t> resolution)
+{
+    VASurfaceAttrib* attribList = NULL;
+    uint32_t numAttribs = 0;
+    VASurfaceID render_surface = VA_INVALID_ID;
+
+    // when ConfigAttribs were not queried just do YUV420 as it is considered
+    // the universal supported format by the driver.  RT formats depend on
+    // profile and entrypoint.
+    unsigned int formats = VA_RT_FORMAT_YUV420;
+
+    if (!m_querySurfaceAttribList.empty()) {
+        numAttribs = m_querySurfaceAttribList.size();
+        attribList = &m_querySurfaceAttribList[0];
+    }
+
+    if (!m_queryConfigAttribList.empty()) {
+        std::vector<VAConfigAttrib>::iterator it = std::find_if(
+            m_queryConfigAttribList.begin(), m_queryConfigAttribList.end(),
+            std::bind(isConfigAttribInList, std::placeholders::_1,
+                      VAConfigAttribRTFormat));
+        formats = it->value;
```

Comment:
bug if `it == m_queryConfigAttribList.end()`.  Please revise this section... no need to check if list is empty first... rather you only need to check if the search found something.

Reply:
Agreed. Updated this part.